### PR TITLE
Fix automated tests.

### PR DESCRIPTION
**Pull request by yaoyuannnn**

### CODE REVIEW

**Comment by yaoyuannnn** [comment]
> Won't this fail as soon as the first test fails? It would be nice if we could run all the tests and look at all the failures at once.

Right. Submitted a new commit to allow running all the tests even if any fails.

The Makefile syntax is really scary. Took me half a day to make this small thing work. I guess that's why people opted for newer auto build tools like bazel and scons.

File: make/Makefile.native [hunk 1]
```diff
@@ -69,11 +70,20 @@ test_bin: $(TEST_BIN)
...
 
-run-tests: $(TEST_BIN) $(BUILD_PY_TESTS)
+run-tests:
 	$(MAKE) -f make/Makefile.native --no-print-directory tests
 	$(MAKE) -f make/Makefile.native --no-print-directory exec
-	@$(foreach t, $^, $(eval RET_VALS += $(shell echo $(t) 1>&2; ./$(t) 1>&2; echo $$?)))
-	@$(if $(filter-out 0, $(RET_VALS)), exit 1)
+	@HAS_ERROR=0;				\
+	for t in $(ALL_TESTS); do 		\
+		echo $$t; ./$$t;		\
```
Comment (xyzsam) [inline make/Makefile.native:78]: This looks much better :) One thing I might add is to pipe stdout/stderr to a temporary file and echo it only when an error occurs. That way it is easier to identify the failing tests.

File: make/Makefile.native [hunk 2]
```diff
@@ -61,26 +61,32 @@ TEST_BIN = $(patsubst %.cpp, %, $(BUILD_TESTS))
...
 run-tests:
-	$(MAKE) -f make/Makefile.native --no-print-directory tests
-	$(MAKE) -f make/Makefile.native --no-print-directory exec
+	@$(MAKE) -f make/Makefile.native --no-print-directory tests
+	@$(MAKE) -f make/Makefile.native --no-print-directory exec
 	@HAS_ERROR=0;				\
 	for t in $(ALL_TESTS); do 		\
-		echo $$t; ./$$t;		\
+		printf "Running test ($$t)";	\
+		./$$t > tmpout 2>&1;		\
```
Comment (xyzsam) [inline make/Makefile.native:79]: I think you need to put `2>&1` before the first redirect.

File: make/Makefile.native [hunk 3]
```diff
@@ -61,26 +61,32 @@ TEST_BIN = $(patsubst %.cpp, %, $(BUILD_TESTS))
...
 		RET_VAL=$$(echo $$?);		\
 		if [ $$RET_VAL -ne 0 ]; then 	\
+			echo "... failed";	\
+			cat tmpout;		\
 			HAS_ERROR=1;		\
+		else				\
+			echo " ... ok";		\
 		fi;				\
 	done;					\
+	rm tmpout;				\
```
Comment (xyzsam) [inline make/Makefile.native:89]: `rm -f` to prevent command failure if tmpout doesn't exist.

**Pull request merged**